DependencyInjection/Compiler/FirewallCompilerPass.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+namespace Scheb\TwoFactorBundle\DependencyInjection\Compiler;
+
+use Ivory\CKEditorBundle\Exception\DependencyInjectionException;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\DefinitionDecorator;
+use Symfony\Component\DependencyInjection\Reference;
+
+class FirewallCompilerPass implements CompilerPassInterface
+{
+    public function process(ContainerBuilder $container)
+    {
+        $authManagerDefinition = $container->getDefinition('security.authentication.manager');
+        $authProviders = $authManagerDefinition->getArgument(0);
+        $firewallServiceIds = $this->getServicesToDecorate($authProviders);
+
+        foreach ($firewallServiceIds as $firewall => $serviceIds) {
+            foreach ($serviceIds as $serviceId) {
+                $decoratedServiceId = $serviceId . '.two_factor_decorator';
+                $container
+                    ->setDefinition($decoratedServiceId, new DefinitionDecorator('scheb_two_factor.security.authentication.provider.decorator'))
```

Comment:
Using a DefinitionDecorator breaks your symfony 4 compatibility. This class has been deprecated in 3.x and remove in 4.0. It has been renamed to `ChildDefinition` because it has to do with decoration, it just extends a definition. I highly recommend you to drop compatibility for symfony <3.3 (lowest maintained 3.x version) for your next minor release (the one that include that changes).

Reply:
Yes, definitely. That prototype has been implemented a while ago, when Symfony4 wasn't a thing yet ;)